errors/error.go

```diff
@@ -0,0 +1,373 @@
+// Copyright 2022 Mineiros GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package errors implements the Terramate standard error type.
+// It's heavily influenced by Rob Pike `errors` package in the Upspin project:
+// 	https://commandcenter.blogspot.com/2017/12/error-handling-in-upspin.html
+package errors
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/hcl/v2"
+)
+
+// Error is the default Terramate error type.
+// It may contains some fields unset, but not all.
+// See E() for its usage.
+type Error struct {
+	// Kind is the kind of error.
+	Kind Kind
+
+	// Description of the error.
+	Description string
+
+	// FileRange holds the error source.
+	FileRange hcl.Range
+
+	// Stack which originated the error.
+	Stack StackMeta
+
+	// Err represents the underlying error.
+	Err error
+}
+
+type (
+	// Kind defines the kind of error.
+	Kind string
+	// StackMeta has the metadata of the stack which originated the error.
+	// Same interface as stack.Metadata.
```

Comment:
```suggestion
	// StackMeta provides metadata about the stack that originated the error.
```

thinking if it makes sense to explicit document that the interface is copied :thinking: 

Also thinking if StackMetadata would not be more clear (could be just me more used to always referencing it as metadata not only as meta).

Reply:
I think it's important to add this hint because changes to this type must be aligned with the definition of stack metadata in the stack package.
Now thinking about this I also would say that we should add a similar hint in the stack package like:

```
// Metadata has all metadata loaded per stack
// Hint: changes to this interface must be replicated in errors.StackMeta.
```

What do you think?

Reply:
Hmm I disagree. Any incompatibilties on the interfaces will be caught at build time already, IMO it only clobbers the docs (imagine a third interface somewhere else, now we have to add comments on all places about all other places). But if you want to do it Im OK, not a strong opinion or anything.

Reply:
> Any incompatibilties on the interfaces will be caught at build time already
how? errors.E() receives interface{}...

Reply:
ah true.. still it seems odd to keep documenting this and hoping it will be enough to avoid issues, for a scenario like that it would be safer to just define a single interface/type and use it on all places then... but that would introduce some sort of types package.